More explicit documentation for release process. (#2185)
sourcegraph/sourcegraph: doc/dev/releases.md

# Releases

This document describes how we release Sourcegraph.

## Goal

The goal of our release process is to make releases boring, regular, and eventually, automated.

## Releases are monthly

> We release Sourcegraph by 10am PT on 20th day of each month. No exception is made for weekends or holidays.

Reviewer: Shipping on Weekends, or even Fridays, limits our ability to follow up when something goes wrong. While I agree that this should be a rare occurrence, I feel that releasing early in the week would be less risky.

What are the trade-offs of releasing on the third Tuesday of each month?

Author: I actually [proposed this](https://github.com/sourcegraph/sourcegraph/pull/2096) earlier and you can read the conversation there.

Clearly we don't want to plan on doing work on holidays and weekends, so the wording here is intentionally `*by* 10am PT on the 20th` instead of `*at* 10am PT on the 20th`.

If you read through the release captain tracking issue checklist you will notice a few things:
1. Dates are defined as "working days" which means even if the 20th is on a Sunday, the final release candidate is tagged on Friday (and the first release candidate is tagged/tested on Tuesday).
2. The first step/duty of the release captain that is documented in the tracking issue template is to decide on the exact schedule for the release. This is important because the exact dates/times that things will happen depends on which timezone that the release captain is in. The release captain has some discretion here on the exact timetable for a given release (I can document this more clearly).

I agree that "we release on the 20th" appears to be riskier by some amount compared to "we release on the third Tuesday". I have three responses to this:
- This perceived risk acts as back-pressure to the engineering team. If everyone knows that we release on weekends and holidays, and if everyone is on the rotation to be release captain, then it becomes that much easier for each teammate to make the decision "it is not worth trying to squeeze this in before the release because I don't want to deal with issues on the weekend or holiday".
- To the extent that releasing on the 20th is riskier...
    - This risk can/should be mitigated by automation in the future (GitLab has this kind of automation).
    - This risk has to be balanced against the benefits of having a fixed release date for customers.

This can be a topic of discussion during team meeting on Monday.

"Release" means:
- The Docker images are available for download.

> - The blog post is published.

Author: It is part of the release process, but is not necessarily time sensitive, and isn't something that an engineer needs to do, so I omitted it from here for now.

> - The blog post is published.

Reviewer 2: Publishing the blog post is also not something engineers will typically do, so I would argue either that should be here or "blog post is published" should not be here.

> - The blog post is published.

Author: I answered the original question imprecisely. This is just the definition of what "released" means from a customer perspective, independent of who is responsible.

As a customer, clearly I need to be able to download the image, and I also expect to be able to find out what has changed (i.e. the blog post).

As a customer, I would not consider the product "unreleased" if I don't see a Tweet or other social media posts.

The release always ships on time, even if it's missing features or bug fixes we hoped to get in ([why?](https://about.gitlab.com/2015/12/07/why-we-shift-objectives-and-not-release-dates-at-gitlab/)).

### Why the 20th?

> We don't want to ship a release too late in December because Sourcegraph's offices are closed December 24 through January 1.

Author: ```suggestion
We don't want to ship a release too late in December because all Sourcegraph employees have a scheduled break December 24 through January 1.
```

> We don't want to ship a release too late in December because Sourcegraph's offices are closed December 24 through January 1.

Reviewer 3: "FTTs" please!

> We don't want to ship a release too late in December because Sourcegraph's offices are closed December 24 through January 1.

Author: I think that "FTTs" is an opaque acronym that we should stop using. "Full time teammates" is fine, but also a mouthful. How about just "the Sourcegraph team" in this context?

> We don't want to ship a release too late in December because Sourcegraph's offices are closed December 24 through January 1.

Reviewer 3: Actually, yeah, FTT would be inappropriate here because this process applies to part-time teammates as well.

"The Sourcegraph team" is fine here, but in general "teammates" is a good, simple alternative.

### Why aren't releases continuous?

Although [Sourcegraph.com](https://sourcegraph.com) is continuously deployed (from sourcegraph/sourcegraph@`master`), the version of Sourcegraph that customers use is not continuously released or updated. This is because:

- We don't think customers would be comfortable with a continuously updated service running on their own infrastructure, for security and stability reasons.
- We haven't built the automated testing and update infrastructure to make continuous customer releases reliable.

In the future, we may introduce continuous releases if these issues become surmountable.

## Release process

What is the process we follow to release?

### Release captains

The release captain is _responsible_ for managing the release process and ensuring that the release happens on time. The release captain may _delegate_ work to other teammates, but such delegation does not absolve the release captain of their responsibility to ensure that delegated work gets done.

> No later than 5 _working days_ before the release day the release captain creates a tracking issue using the [release issue template](release_issue_template.md) and assigns it to themself to complete.

Reviewer: TIL: themself

(Only knew *themselves*)

Author: This history of the word is quite interesting.
https://www.merriam-webster.com/words-at-play/themself

| Version | Captain | Release Date |
|---------|---------|--------------|
| 3.0 | @nicksnyder | 2019-02-07 (Wednesday) |
| 3.1 | @nicksnyder | 2019-02-20 (Wednesday) |
| 3.2 | @nicksnyder | 2019-03-20 (Wednesday) |
| 3.3 | @beyang | 2019-04-20 (Saturday) |
| 3.4 | @ggilmore | 2019-05-20 (Monday) |
| 3.5 | @keegancsmith | 2019-06-20 (Thursday) |
| 3.6 | @slimsag | 2019-07-20 (Saturday) |
| 3.7 | @ijsnow | 2019-08-20 (Tuesday) |
| 3.8 | @tsenart | 2019-09-20 (Friday) |
| 3.9 | @lguychard | 2019-10-20 (Sunday) |
| 3.10 | @attfarhan | 2019-11-20 (Wednesday) |
| 3.11 | @chrismwendt | 2019-12-20 (Saturday) |
| 3.12 | @vanesa | 2020-01-20 (Monday, MLK) |
| 3.13 | @felixfbecker | 2020-02-20 (Thursday) |

### Release branches

> Each release of [Sourcegraph](https://github.com/sourcegraph/sourcegraph) is tagged from a release branch.

Reviewer: Do we have our versioning scheme documented somewhere? How do we increment Major, Minor and Patch versions?

Author: Does https://github.com/sourcegraph/sourcegraph/pull/2185/commits/726794c5fa8a6a248e27d23c7201b921ba05b063 answer your question? If not, can you clarify what question you would have if you were the release captain for the next release?

1. The release captain (different person) for 3.1 creates the `3.1` release branch at commit `H` and a new release cycle begins.
1. Commit `J` fixes a critical bug that impacts both 3.0 and 3.1, so it is cherry picked into both `3.0` and `3.1` release branches and new releases are tagged (`v3.0.2`, `v3.1.2`).

```
A---B---C---D---E---F---G---H---I---J---K---L (master branch)
     \                       \                   
      \                       `---v3.1.0-rc.1---I'---v3.1.0---J'---v3.1.2 (3.1 release branch)
       \
        `---v3.0.0-rc.1---D'---v3.0.0---F'---v3.0.1---J'---v3.0.2 (3.0 release branch)
```

### Issues

How do we deal with issues that are found during the release process?

#### Blocking

The release always ships on time, even if it's missing features or bug fixes we hoped to get in ([why?](https://about.gitlab.com/2015/12/07/why-we-shift-objectives-and-not-release-dates-at-gitlab/)).

There are only two kinds of issues that are eligible to be labeled as "release-blocking".
1. Issues that literally prevent us from tagging a release (i.e. our CI logic to produce builds from git tags is broken).

> 2. Issues that break the critical workflows explicitly tested in the [release issue template](release_issue_template.md) which would impact _most_ customers. This means that most customers would find our product fundamentally broken.

Reviewer: What about critical security issues?

Author: Add exception for critical security regressions.

Only the release captain can label something as release blocking.

The release captain has unlimited power to make changes to the release branch to resolve release blocking issues, including but not limited to:
- Fixing the issue in the release branch.
- Reverting commits in the release branch.
- Re-cutting the release branch from an older working commit off master.

#### Non-blocking

> Most issues are non-blocking. Fixes to non-blocking issues can be fixed in `master` by the code owner and then `git cherry-pick` into the release branch with the approval of the release captain. Alternatively, broken features can be reverted out of the release branch or disabled via feature flag if they aren't ready or too buggy.

Reviewer: ```suggestion
Most issues are non-blocking. Fixes to non-blocking issues can be fixed in `master` by the code owner who can then `git cherry-pick` those commits into the release branch with the approval of the release captain. Alternatively, broken features can be reverted out of the release branch or disabled via feature flags if they aren't ready or are too buggy.
```

# Releases

This document describes how we release Sourcegraph.

## Goal

The goal of our release process is to make releases boring, regular, and eventually, automated.

## Releases are monthly

> We release Sourcegraph by 10am PT on the 20th day of each month. No exception is made for weekends or holidays.

Reviewer 4: I was going to add this during the meeting but there were more it's more of a nuetral/non-argument and I didn't want to interrupt the discussion: 

If the work of the release captain is to do all of the steps "on or before" the release date, and that means it would be fine to do all the final work for the release on the 18th if the 20th is a monday AND a holiday, then there's not real difference between whether we release on the 20th or the 3rd tuesday. The fact that our end goal is for the release is to be a "non-event" and ideally automated away in the future also supports this opinion. Since it should be a "non-event" and the work can be done before the release date in the mean time, I think we should go with the option that takes less thought, which is IMO releasing on a fixed date (though the difference is pretty small).

Author: Thanks for sharing your thoughts!

After re-reading this, I think the "no exception is made for weekends or holidays" is somewhat redundant so I am going to remove it.

"Release" means:
- The Docker images are available for download.
- The blog post is published.

The release always ships on time, even if it's missing features or bug fixes we hoped to get in ([why?](https://about.gitlab.com/2015/12/07/why-we-shift-objectives-and-not-release-dates-at-gitlab/)).

### Why the 20th?

We don't want to ship a release too late in December because the Sourcegraph team has a scheduled break December 24 through January 1.

### Why aren't releases continuous?

Although [Sourcegraph.com](https://sourcegraph.com) is continuously deployed (from sourcegraph/sourcegraph@`master`), the version of Sourcegraph that customers use is not continuously released or updated. This is because:

- We don't think customers would be comfortable with a continuously updated service running on their own infrastructure, for security and stability reasons.
- We haven't built the automated testing and update infrastructure to make continuous customer releases reliable.

In the future, we may introduce continuous releases if these issues become surmountable.

## Versioning

> [Monthly releases](#releases-are-monthly) of Sourcegraph increase the minor version number (e.g. 3.1 -> 3.2). These releases do not require any manual migration steps.

Reviewer 5: ```suggestion
[Monthly releases](#releases-are-monthly) of Sourcegraph increase the minor version number (e.g. 3.1 -> 3.2). These releases **never** require any manual migration steps.
```

> Patch releases (e.g. 3.0.0 -> 3.0.1) are released on an as-needed basis to fix bugs and security issues. These releases do not require any manual migration steps.

Reviewer 2: ```suggestion
Patch releases (e.g. 3.0.0 -> 3.0.1) are released on an as-needed basis to fix bugs and security issues. These releases **never** require any manual migration steps.
```

Reviewer 2: I suggest this better wording + emphasis because I thought you also said below major version numbers do not require manual migrations and was surprised by this. It turned out I just misread, this would've prevented me from misreading.

> On rare occasions we may decide to increase the major version number (e.g. 2.13 -> 3.0). These releases may require manual migration steps.

Reviewer 2: ```suggestion
On rare occasions we may decide to increase the major version number (e.g. 2.13 -> 3.0). These releases **may** require manual migration steps.
```

Reviewer: +1

### Release captains

The release captain is _responsible_ for managing the release process and ensuring that the release happens on time. The release captain may _delegate_ work to other teammates, but such delegation does not absolve the release captain of their responsibility to ensure that delegated work gets done.

No later than 5 _working days_ before the release day the release captain creates a tracking issue using the [release issue template](release_issue_template.md) and assigns it to themself to complete.

| Version | Captain | Release Date |
|---------|---------|--------------|
| 3.0 | @nicksnyder | 2019-02-07 (Wednesday) |
| 3.1 | @nicksnyder | 2019-02-20 (Wednesday) |
| 3.2 | @nicksnyder | 2019-03-20 (Wednesday) |
| 3.3 | @beyang | 2019-04-20 (Saturday) |
| 3.4 | @ggilmore | 2019-05-20 (Monday) |
| 3.5 | @keegancsmith | 2019-06-20 (Thursday) |
| 3.6 | @slimsag | 2019-07-20 (Saturday) |
| 3.7 | @ijsnow | 2019-08-20 (Tuesday) |
| 3.8 | @tsenart | 2019-09-20 (Friday) |
| 3.9 | @lguychard | 2019-10-20 (Sunday) |
| 3.10 | @attfarhan | 2019-11-20 (Wednesday) |
| 3.11 | @chrismwendt | 2019-12-20 (Saturday) |
| 3.12 | @vanesa | 2020-01-20 (Monday, MLK) |

> | 3.13 | @felixfbecker | 2020-02-20 (Thursday) |

Reviewer 2: This section should describe what we do when a release captain is unexpectedly not available, or if they need to plan to be unavailable during their assigned versions.

### Release branches

Each major and minor release of [Sourcegraph](https://github.com/sourcegraph/sourcegraph) has a long lived release branch (e.g. `3.0`, `3.1`). Individual releases are tagged from these release branches (e.g. `v3.0.0-rc.1`, `v3.0.0`, `v3.0.1-rc.1`, and `v3.0.1` would be tagged from the `3.0` release branch).

To avoid confusion between tags and branches:

- Tags are always the full semantic version with a leading `v` (e.g. `v2.10.0`)
- Branches are always the dot-separated major/minor versions with no leading `v` (e.g. `2.10`).

> Development always happens on master `master` and changes are cherry picked onto release branch as necessary with the approval of the release captain.

Reviewer 2: ```suggestion
Development always happens on `master` and changes are cherry picked onto release branch as necessary with the approval of the release captain.
```

1. The release captain (different person) for 3.1 creates the `3.1` release branch at commit `H` and a new release cycle begins.
1. Commit `J` fixes a critical bug that impacts both 3.0 and 3.1, so it is cherry picked into both `3.0` and `3.1` release branches and new releases are tagged (`v3.0.2`, `v3.1.2`).

```
A---B---C---D---E---F---G---H---I---J---K---L (master branch)
     \                       \                   
      \                       `---v3.1.0-rc.1---I'---v3.1.0---J'---v3.1.2 (3.1 release branch)
       \
        `---v3.0.0-rc.1---D'---v3.0.0---F'---v3.0.1---J'---v3.0.2 (3.0 release branch)
```

### Issues

How do we deal with issues that are found during the release process?

#### Blocking

The release always ships on time, even if it's missing features or bug fixes we hoped to get in ([why?](https://about.gitlab.com/2015/12/07/why-we-shift-objectives-and-not-release-dates-at-gitlab/)).

There are only three kinds of issues that are eligible to block a release:

1. Issues that literally prevent us from tagging a release (i.e. our CI logic to produce builds from git tags is broken).

> 2. Issues that break one of the critical workflows explicitly tested in the [release issue template](release_issue_template.md) which would impact _most_ customers and don't have workarounds. 

Reviewer 2: Either we make exceptions to this rule regularly, or we need to ensure the release issue template is very exhaustive in what we test. For example, some things that we would apparently consider non-blocking issues and ship anyway:

- Monaco editor on the site config page is broken, literally nobody can configure their instance.
- Non-security regression in management console which locks all site admins out when using that Sourcegraph version.
- Non-security regression in auth which locks 50% of all our customers user's out of their Sourcegraph instance (e.g. anyone using github SSO is broken).
- Bad CSS change which entirely breaks display of the main repo page on sourcegraph instances.

etc. The above is just a few exceptions I can think of. I realize I am being pedantic, but we are trying to formalize this process so we should be clear that -- yeah -- sometimes we won't be 100% accurate in what we test and sometimes there will be issues which e.g. the PM declares a release-blocker.

Author: > make exceptions to this rule regularly

The goal is to have a rule that is clear and we never make exceptions for.

I have updated the language to be more general here, and to clarify that the default resolution to all release blocking issues is to revert back to a known good state in the release branch.
